Tests/MongoSwiftTests/DocumentTests.swift

```diff
@@ -372,4 +373,101 @@ final class DocumentTests: XCTestCase {
         expect(doc["a1"]).to(equal(arr1))
         expect(doc["a2"]).to(equal(arr2))
     }
+
+    func testSetValueForExistingKey() throws {
+        let overwritables: Document = ["double": 2.5, "int32": Int32(32), "int64": Int64(64), "bool": false]
+        let oid = ObjectId()
+        let nonOverwritables: Document = ["string": "hello", "nil": nil, "doc": ["x": 1] as Document,
+                                          "arr": [1, 2] as [Int], "oid": oid]
+
+        // test replacing `Overwritable` types with values of their own type
+        do {
+            var doc = overwritables
+
+            // overwrite int32 with int32
+            doc["int32"] = Int32(15)
+            expect(doc["int32"] as? Int).to(equal(15))
+
+            // overwrite int32 with int
+            doc["int32"] = 20
+            expect(doc["int32"] as? Int).to(equal(20))
+
+            // overwrite int64 with int64
+            let bigInt64 = Int64(Int32.max) + 1
+            doc["int64"] = bigInt64
+
+            let bigInt = Int(bigInt64) + 1
+            doc["int64"] = bigInt
+
+            doc["bool"] = true
+            doc["double"] = 3.0
+
+            // final values
+            expect(doc).to(equal(["double": 3.0, "int32": 20, "int64": bigInt, "bool": true]))
+        }
+
+        // test replacing some of the non-Overwritable types with values of their own types
+        do {
+            var doc = nonOverwritables
+
+            doc["string"] = "hi"
+            doc["nil"] = nil
```

Comment:
Does it make sense to optimize for the `nil` case? BSON types with no custom data (e.g. nil, undefined, MinKey, MaxKey) should be overwritable as a no-op. That would allow us to avoid creating a new document and re-copying all other keys.